File: 프로그래머스/2/150368. 이모티콘 할인행사/이모티콘 할인행사.py
import heapq

def solution(users, emoticons):
    answer = [0, 0]
    rates = [ 0 for _ in range(len(emoticons)) ]
    queue = []
    
    def calculate(rates):
        result = [0, 0, 0, 0]
        count = 0
        sumPrice = 0
        for u in users:
            price = 0
            for i in range(len(emoticons)):
                if rates[i] >= u[0]:
                    price += int(emoticons[i] * (100 - rates[i]) / 100)
                    if price >= u[1]:
                        result[2] += 1
                        price = 0
                        break
            result[3] += price
        
        if result[2] > 0:
            result[0] = result[2] * -1
        if result[3] > 0:
            result[1] = result[3] * -1
        return result
                
    
    def dfs(l):
        if l == len(emoticons):
            heapq.heappush(queue, calculate(rates))
            return
        
        for rate in range(10, 50, 10):
            rates[l] = rate
            dfs(l+1)
    dfs(0)
    
    answer[0] = queue[0][2]
    answer[1] = queue[0][3]
    
    return answer

File: 프로그래머스/2/150368. 이모티콘 할인행사/이모티콘 할인행사.kt
import java.util.*

class Solution {
    fun solution(users: Array<IntArray>, emoticons: IntArray): IntArray {
        var answer: IntArray = intArrayOf()
        var rates = MutableList(emoticons.size) { 0 }
        var queue = PriorityQueue(compareByDescending<IntArray> { it[0] }.thenByDescending { it[1] })
        
        fun findSolution(): IntArray {
            val result = intArrayOf(0, 0)
            var count = 0;
            var sum = 0
            userLoop@ for(u in users) {
                var price = 0;
                for(i in emoticons.indices) {
                    if(rates[i] >= u[0]) {
                        price += emoticons[i] * (100-rates[i]) / 100
                        if(price >= u[1]) {
                            result[0]++;
                            count++;
                            continue@userLoop
                        }
                    }
                }
                result[1] += price
                sum += price
            }
            
            return intArrayOf(count, sum)
        }
        
        fun DFS(L: Int) {
            if(L == emoticons.size) {
                queue.offer(findSolution())
            }   
            else {
                for(i in 10..40 step (10)) {
                    rates[L] = i;
                    DFS(L+1);
                }
            }
        }
        
        DFS(0)

        return queue.peek()
    }
}